lib/elsa/dynamic_process_manager.ex

```diff
@@ -0,0 +1,140 @@
+    GenServer.start_link(__MODULE__, init_arg, server_opts)
+  end
+
+  def child_spec(init_arg) do
+    id = Keyword.fetch!(init_arg, :id)
+
+    %{
+      id: id,
+      start: {__MODULE__, :start_link, [init_arg]}
+    }
+  end
+
+  def ready?(server, timeout \\ 10_000) do
+    GenServer.call(server, :ready?, timeout)
+  end
+
+  def init(init_arg) do
+    Process.flag(:trap_exit, true)
+
+    dynamic_supervisor = Keyword.fetch!(init_arg, :dynamic_supervisor)
+    synchronous? = Keyword.get(init_arg, :synchronous, false)
+
+    state = %{
+      dynamic_supervisor: dynamic_supervisor,
+      dynamic_supervisor_ref: whereis(dynamic_supervisor) |> Process.monitor(),
+      child_specs: Keyword.get(init_arg, :children, [])
+    }
+
+    initializer =
+      case Keyword.get(init_arg, :initializer) do
+        {module, function, args} -> fn -> apply(module, function, args) end
+        function when is_function(function, 0) -> function
+        _ -> fn -> [] end
+      end
+
+    case synchronous? do
+      false ->
+        {:ok, state, {:continue, {:initialize, initializer}}}
+
+      true ->
+        {:noreply, new_state} = handle_continue({:initialize, initializer}, state)
+        {:ok, new_state}
+    end
+  end
+
+  def handle_continue({:initialize, initializer}, state) do
+    new_state =
+      Map.update!(state, :child_specs, fn specs ->
+        specs ++ run_initializer(initializer)
+      end)
+
+    start_children(new_state)
+    {:noreply, new_state}
+  end
+
+  def handle_call({:start_child, child}, _from, state) do
+    output = DynamicSupervisor.start_child(state.dynamic_supervisor, child)
+    {:reply, output, Map.update!(state, :child_specs, fn specs -> specs ++ [child] end)}
+  end
+
+  def handle_call(:ready?, _from, state) do
+    {:reply, true, state}
+  end
+
+  def handle_info({:DOWN, ref, _, _, _}, %{dynamic_supervisor_ref: ref} = state) do
+    Logger.info(fn ->
+      "#{__MODULE__}: Dynamic Supervisor #{state.dynamic_supervisor} has died, restarting recorded children"
+    end)
+
+    wait_for(state.dynamic_supervisor)
+
+    start_children(state)
+    {:noreply, state}
+  end
+
+  defp start_children(state) do
+    state.child_specs
+    |> Enum.each(fn child ->
+      output = DynamicSupervisor.start_child(state.dynamic_supervisor, child)
+
+      Logger.debug(fn ->
+        "#{__MODULE__}: output of starting #{inspect(child)}: #{inspect(output)}"
+      end)
+    end)
+  end
+
+  defp run_initializer(initializer) do
+    initializer.()
+  rescue
+    e ->
+      Logger.warn(fn -> "#{__MODULE__}: initializer raised exception, retrying: #{inspect(e)}" end)
+
+      Process.sleep(1_000)
+      run_initializer(initializer)
+  catch
+    :exit, e ->
+      Logger.warn(fn -> "#{__MODULE__}: initializer raised exception, retrying: #{inspect(e)}" end)
+
+      Process.sleep(1_000)
+      run_initializer(initializer)
+  end
+
+  defp whereis(name) when is_atom(name) do
+    Process.whereis(name)
```

Comment:
if you're going to pull this into a private function here (just to add the atom guard on the supervisor name?) would it make sense to just fold in the `Process.monitor/1` and return the ref directly? Conversely, would it be better to bubble up higher to the public API that you're expecting to only accept an atom (module name) for your supervisor?